File: Course.java
import java.util.ArrayList;

/**
* This class is designed to represent a course from the course catalog.
*/
public class Course{

  private String name; //Holds the name of the course
  private String[] tags; //Holds the tags associated with the course
  private int[] timeslots; //Holds the timeslots of the course parsed into integers.
  private String[] preReqs; //Holds the list of preReqs of the course.

  /**
   * Construct a Course instance.
   *
   * @pre name, tags, timeslots, and preReqs are all valid and of the correct type.
   * @post constructs a Course instance with the given input.
   *
   * @param name Name of Course.
   * @param tags Tags associated to the Course.
   * @param timeslotsString Unparsed timeslots associated to the Course.
   * @param preReqs Pre-requisite courses required by the Course.
   */
  public Course(String name, String[] tags, String[] timeslotsString, String[] preReqs){
    this.name = name;
    this.tags = tags;
    this.timeslots = parseTimeSlots(timeslotsString);
    this.preReqs = preReqs;
  }

  /**
   * Parses the raw timeslots into an array of integers.
   *
   * @post returns an array of integers which represent timeslots of the course.
   *
   * @param toBeParsed the string array representing all the timeslots of the course.
   * @return an array of integers which represent timeslots of the course.
   */
  private int[] parseTimeSlots(String[] toBeParsed){
    ArrayList<Integer> result = new ArrayList<Integer>();

    for(int i = 0; i < toBeParsed.length; i++){
        String days = toBeParsed[i].split(" ", 2)[0];
        //This has to be parsable - edge cases might happen
        //TODO DEAL WITH EDGE CASES
        String[] times = toBeParsed[i].split(" ", 2)[1].split(" - ");

        for(int j = 0; j < days.length(); j++){
          char day = days.charAt(j);
          int dayToInt;
          switch(day){
            case 'M':
              dayToInt = 0;
              break;
            case 'T':
              dayToInt = 1440;
              break;
            case 'W':
              dayToInt = 1440 * 2;
              break;
            case 'R':
              dayToInt = 1440 * 3;
              break;
            case 'F':
              dayToInt = 1440 * 4;
              break;
            default:
              dayToInt = 0;
              break;
          }

          result.add(dayToInt + parseTime(times[0]));
          result.add(dayToInt + parseTime(times[1]));
        }
    }

    //Converting ArrayList to primative Array
    int[] resultArray = new int[result.size()];

    for(int i = 0; i < result.size(); i++){
      resultArray[i] = result.get(i);
    }

    return resultArray;
  }

  /**
   * Parses the time from hh:ss format to an integer in terms of number
   * of minutes out the day.
   *
   * @post returns the integer representation of the time
   *
   * @return the integer representation of time
   */
  private int parseTime(String time){
    String[] hoursAndMins = time.split(":");
    return Integer.parseInt(hoursAndMins[0]) * 60 +Integer.parseInt(hoursAndMins[1]);
  }

  /**
   * Fetch the name of the Course.
   *
   * @post returns the name variable
   *
   * @return name
   */
  public String getName(){
    return name;
  }


  /**
   * Fetch the list of tags assoicated to the Course.
   *
   * @post returns tags
   *
   * @return tags
   */
  public String[] getTags(){
    return tags;
  }


  /**
   * Fetch the timeslots that the Course takes.
   *
   * @post returns timeslots
   *
   * @return timeslots
   */
  public int[] getTimeslots(){
    return timeslots;
  }


  /**
   * Fetch the pre-requisites of the Course.
   *
   * @post returns preReqs
   *
   * @return preReqs
   */
  public String[] getPreReqs(){
    return preReqs;
  }
}

File: javascript/Course.js

/**
 * Construct a Course instance.
 *
 * @pre name, tags, timeslots, and preReqs are all valid and of the correct type.
 * @post constructs a Course instance with the given input.
 *
 * @param {string} name Name of Course.
 * @param {Array} tags Tags associated to the Course.
 * @param {Array} timeslotsString Unparsed timeslots associated to the Course.
 * @param {Array} preReqs Pre-requisite courses required by the Course.
 * @class
 */
var Course = (function () {
    function Course(name, tags, timeslotsString, preReqs) {
     //  if (this.subject === undefined)
    	//     this.subject = null;
     //  if (this.number === undefined)
    	//     this.number = null;
    	// if (this.attributes === undefined)
    	//     this.attributes = null;
    	if (this.name === undefined)
    	    this.name = null;
      // if (this.prof === undefined)
    	 //    this.prof = null;
      // if (this.description === undefined)
    	 //    this.description = null;
    	if (this.tags === undefined)
    	    this.tags = null;
    	if (this.timeslots === undefined)
    	    this.timeslots = null;
    	if (this.preReqs === undefined)
    	    this.preReqs = null;
      // this.subject = subject;
      // this.number = number;
      // this.attributes = attributes.split(",");
    	this.name = name;
      // this.prof = prof;
      // this.description = description;
    	this.tags = tags;
      this.timeslots = this.parseTimeSlots(timeslotsString);
    	// this.timeslots =  this.parseTimeSlots(timeslotsString);
    	this.preReqs = preReqs;
//      console.log(this.attributes);
    }
   /**
     * Parses the raw timeslots into an array of integers.
     *
     * @post returns an array of integers which represent timeslots of the course.
     *
     * @param {Array} toBeParsed the string array representing all the timeslots of the course.
     * @return {Array} an array of integers which represent timeslots of the course.
     * @private
     */

    Course.prototype.parseTimeSlots = function (toBeParsed) {
      //TODO make it less hacky.
    	var result = ([]);

    	for (var i = 0; i < toBeParsed.length; i++) {
    	    {

        //Get rid of numbers to parse out days.
        var days = toBeParsed[i].split(/([0-9]+)/)[0];

    		var timeArr = toBeParsed[i].split("-");

        //Get rid of letters to parse out times. Obtain starting time.
        var startTime = timeArr[0].split(/[a-zA-Z]+/)[1];

        startTime = startTime.replace(/ +/g,'');

        //Obtain ending time.
        var endTime = timeArr[1];

        //Array to hold starting and ending times.
        var times = new Array();
        times.push(startTime);
        times.push(endTime);

    		for (var j = 0; j < days.length; j++) {
          //Takes string containing days of the week and iterates through each day.
          var day = days.charAt(j);
    			var dayToInt = void 0;

          //Looks at UTF-16 representations of days of the week.
    			switch (day.charCodeAt(0)) {
    			case 77 /* 'M' */:
    			    dayToInt = 0;
    			    break;
    			case 84 /* 'T' */:
    			    dayToInt = 1440;
    			    break;
    			case 87 /* 'W' */:
    			    dayToInt = 1440 * 2;
    			    break;
    			case 82 /* 'R' */:
    			    dayToInt = 1440 * 3;
    			    break;
    			case 70 /* 'F' */:
    			    dayToInt = 1440 * 4;
    			    break;
    			default:
    			    dayToInt = 0;
    			    break;
    			}

          if(dayToInt != -1){
          /* add */ (result.push(dayToInt + this.parseTime(times[0])) > 0);
          /* add */ (result.push(dayToInt + this.parseTime(times[1])) > 0);
          }


    		}
    	    };
    	}

    	var resultArray = (function (s){
          var a = new Array();
          console.log(s);
          while (s-- > 0){
            a.push(0);
            //console.log(a);
          }
    	     return a; })(/* size */ result.length);
    	for (var i = 0; i < result.length; i++) resultArray[i] = result[i];

      console.log(resultArray);
    	return resultArray;
    };
        /**
     * Parses the time from hh:ss format to an integer in terms of number
     * of minutes out the day.
     *
     * @post returns the integer representation of the time
     *
     * @return {number} the integer representation of time
     * @param {string} time
     * @private
     */
	Course.prototype.parseTime = function (timeString) {
  var hoursAndMins = timeString.split(":");
  var minutes = parseInt(hoursAndMins[0]) * 60 + parseInt(hoursAndMins[1]);
	return minutes;
    };

    //} //END OF COMMENTED FALSE JUMP STATEMENT


    /**
     * Fetch the name of the Course.
     *
     * @post returns the name variable
     *TODO
     * @return {string} name
     */
    Course.prototype.getSubject = function () {
	return this.subject;
    };

          /**
     * Fetch the name of the Course.
     *
     * @post returns the name variable
     *TODO
     * @return {string} name
     */
    Course.prototype.getNumber = function () {
	return this.number;
    };

          /**
     * Fetch the name of the Course.
     *
     * @post returns the name variable
     *TODO
     * @return {string} name
     */
    Course.prototype.getAttributes = function () {
	return this.attributes;
    };






    /**
     * Fetch the name of the Course.
     *
     * @post returns the name variable
     *
     * @return {string} name
     */
    Course.prototype.getName = function () {
	return this.name;
    };
        /**
     * Fetch the list of tags assoicated to the Course.
     *
     * @post returns tags
     *
     * @return {Array} tags
     */
    Course.prototype.getTags = function () {
	return this.tags;
    };

    //TODO Comments.

    Course.prototype.getTimeslotString = function(){
      return this.timeslotsString;
    };
    Course.prototype.getProf = function(){
      return this.prof;
    };
        Course.prototype.getDescription = function(){
      return this.description;
    };

        /**
     * Fetch the timeslots that the Course takes.
     *
     * @post returns timeslots
     *
     * @return {Array} timeslots
     */
    Course.prototype.getTimeslots = function () {
	return this.timeslots;
    };
        /**
     * Fetch the pre-requisites of the Course.
     *
     * @post returns preReqs
     *
     * @return {Array} preReqs
     */
    Course.prototype.getPreReqs = function () {
	return this.preReqs;
    };
    return Course;
}());
Course["__class"] = "Course";

//End Course.js
